compiler/rustc_mir_build/src/builder/expr/into.rs

```diff
@@ -330,6 +330,10 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                                     );
                                 }
 
+                                // handle patterns like `None | None`
+                                arm_blocks.sort_by_key(|(_, discr, _, _)| discr.val);
+                                arm_blocks.dedup_by_key(|(_, discr, _, _)| discr.val);
```

Comment:
If you have two arms matching the same value this happens to work and pick the first arm I think because this is a stable sort, but it is rather subtle reason why it works.

Reply:
right. I've added a comment noting this. I think that should be fine for now, right?